src/gui/qgsrelationwidget.h

```diff
@@ -0,0 +1,260 @@
+/***************************************************************************
+                         qgsrelationwidget.h
+                         ----------------------
+    begin                : October 2020
+    copyright            : (C) 2020 by Ivan Ivanov
+    email                : ivan@opengis.ch
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSRELATIONWIDGET_H
+#define QGSRELATIONWIDGET_H
+
+#include <QWidget>
+#include <QToolButton>
+#include <QButtonGroup>
+#include <QGridLayout>
+#include "qobjectuniqueptr.h"
+
+#include "qobjectuniqueptr.h"
+#include "qgsattributeeditorcontext.h"
+#include "qgscollapsiblegroupbox.h"
+#include "qgsdualview.h"
+#include "qgsrelation.h"
+#include "qgsvectorlayerselectionmanager.h"
+#include "qgis_gui.h"
+
+class QgsFeature;
+class QgsVectorLayer;
+class QgsVectorLayerTools;
+class QgsMapTool;
+class QgsMapToolDigitizeFeature;
+
+
+/**
+ * Base class to build new relation widgets.
+ * \ingroup gui
+ * \class QgsRelationWidget
+ * \since QGIS 3.18
+ */
+class GUI_EXPORT QgsRelationWidget : public QWidget
+{
+    Q_OBJECT
+
+  public:
+
```

Comment:
This will need sip casting code to convert to the subclasses

Reply:
Done, thanks